diff --git a/stress-ng.c b/stress-ng.c
index fdbe1a2..559aaaa 100644
--- a/stress-ng.c
+++ b/stress-ng.c
@@ -26,6 +26,8 @@

 #include <getopt.h>
 #include <syslog.h>
+#include <pthread.h>
+#include <time.h>

 #if defined(__linux__)
 #include <sys/utsname.h>
@@ -2119,6 +2121,23 @@ static uint32_t get_total_num_procs(proc_info_t *procs_list)
 }

 /*
+ * monitor_counter()
+ * Outputs stressor APMs during run
+ */
+void* monitor_counter(void* counter) {
+    uint64_t* real_counter = (uint64_t*)counter;
+    uint64_t last = 0;
+	while (true) {
+        shim_usleep(1000000);
+        time_t now;
+        time(&now);
+		pr_inf("Time %ld, counter: %" PRIu64 "\n", now, *real_counter-last);
+        last = *real_counter;
+		fflush(stdout);
+	}
+}
+
+/*
  *  stress_run ()
  *	kick off and run stressors
  */
@@ -2217,8 +2236,18 @@ again:
 							.page_size = stress_get_pagesize(),
 						};

+						pthread_t monitor;
+						int prc;
+						prc = pthread_create(&monitor, NULL, monitor_counter, (void*)args.counter);
+						if (prc) {
+							pr_fail("Failed to create monitoring thread.");
+							rc = EXIT_FAILURE;
+							goto child_exit;
+						}
 						rc = proc_current->stressor->stress_func(&args);
 						stats->run_ok = (rc == EXIT_SUCCESS);
+						pthread_cancel(monitor);
+						pthread_join(monitor, NULL);
 					}
 #if defined(STRESS_PERF_STATS)
if (g_opt_flags & OPT_FLAGS_PERF_STATS) {